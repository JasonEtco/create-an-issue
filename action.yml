name: Create an issue
description: Creates a new issue using a template with front matter.
runs:
  using: node20
  main: dist/index.js
branding:
  icon: alert-circle
  color: gray-dark
inputs:
  assignees:
    description: GitHub handle of the user(s) to assign the issue (comma-separated)
    required: false
  milestone:
    description: Number of the milestone to assign the issue to
    required: false
  filename:
    description: The name of the file to use as the issue template
    default: .github/ISSUE_TEMPLATE.md
    required: false
  update_existing:
    description: Update an open existing issue with the same title if it exists
    required: false
  search_existing:
    description: Existing types of issues to search for (comma-separated)
    required: false
    default: open
outputs:
  number:
    description: Number of the issue that was created
  url:
    description: URL of the issue that was created
