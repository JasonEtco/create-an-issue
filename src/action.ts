import * as core from "@actions/core";
import { Toolkit } from "actions-toolkit";
import fm from "front-matter";
import nunjucks from "nunjucks";
// @ts-expect-error
import dateFilter from "nunjucks-date-filter";
import { ZodError } from "zod";
import {
  FrontMatterAttributes,
  frontmatterSchema,
  listToArray,
  setOutputs,
} from "./helpers";

function logError(
  tools: Toolkit,
  template: string,
  action: "creating" | "updating" | "parsing",
  err: any
) {
  // Log the error message
  const errorMessage = `An error occurred while ${action} the issue. This might be caused by a malformed issue title, or a typo in the labels or assignees. Check ${template}!`;
  tools.log.error(errorMessage);
  tools.log.error(err);

  // The error might have more details
  if (err.errors) tools.log.error(err.errors);

  // Exit with a failing status
  core.setFailed(errorMessage + "\n\n" + err.message);
  return tools.exit.failure();
}

export async function createAnIssue(tools: Toolkit) {
  const template = tools.inputs.filename || ".github/ISSUE_TEMPLATE.md";
  const assignees = tools.inputs.assignees;

  let updateExisting: Boolean | null = null;
  if (tools.inputs.update_existing) {
    if (tools.inputs.update_existing === "true") {
      updateExisting = true;
    } else if (tools.inputs.update_existing === "false") {
      updateExisting = false;
    } else {
      tools.exit.failure(
        `Invalid value update_existing=${tools.inputs.update_existing}, must be one of true or false`
      );
    }
  }

  const env = nunjucks.configure({ autoescape: false });
  env.addFilter("date", dateFilter);

  let repo: { owner: string, repo: string } = tools.context.repo;
  if (tools.inputs.repo) {
    const [owner, repoName] = tools.inputs.repo.split("/");
    repo = { owner, repo: repoName };
  }
  const templateVariables = {
    ...tools.context,
    repo: repo,
    env: process.env,
    date: Date.now(),
  };

  // Get the file
  tools.log.debug("Reading from file", template);
  const file = (await tools.readFile(template)) as string;

  // Grab the front matter as JSON
  const { attributes: rawAttributes, body } = fm<FrontMatterAttributes>(file);

  let attributes: FrontMatterAttributes;
  try {
    attributes = await frontmatterSchema.parseAsync(rawAttributes);
  } catch (err) {
    if (err instanceof ZodError) {
      const formatted = err.format();
      return logError(tools, template, "parsing", formatted);
    }
    throw err;
  }

  tools.log(`Front matter for ${template} is`, attributes);

  const templated = {
    body: env.renderString(body, templateVariables),
    title: env.renderString(attributes.title, templateVariables),
  };
  tools.log.debug("Templates compiled", templated);

  if (updateExisting !== null) {
    tools.log.info(`Fetching issues with title "${templated.title}"`);

    let query = `is:issue repo:${
      repo.owner + "/" + repo.repo
    } in:title "${templated.title.replace(/['"]/g, "\\$&")}"`;

    const searchExistingType = tools.inputs.search_existing || "open";
    const allowedStates = ["open", "closed"];
    if (allowedStates.includes(searchExistingType)) {
      query += ` is:${searchExistingType}`;
    }

    const existingIssues = await tools.github.search.issuesAndPullRequests({
      q: query,
    });
    const existingIssue = existingIssues.data.items.find(
      (issue) => issue.title === templated.title
    );
    if (existingIssue) {
      if (updateExisting === false) {
        tools.exit.success(
          `Existing issue ${existingIssue.title}#${existingIssue.number}: ${existingIssue.html_url} found but not updated`
        );
      } else {
        try {
          tools.log.info(
            `Updating existing issue ${existingIssue.title}#${existingIssue.number}: ${existingIssue.html_url}`
          );
          const issue = await tools.github.issues.update({
            ...repo,
            issue_number: existingIssue.number,
            body: templated.body,
          });
          setOutputs(tools, issue.data);
          tools.exit.success(
            `Updated issue ${existingIssue.title}#${existingIssue.number}: ${existingIssue.html_url}`
          );
        } catch (err: any) {
          return logError(tools, template, "updating", err);
        }
      }
    } else {
      tools.log.info("No existing issue found to update");
    }
  }

  // Create the new issue
  tools.log.info(`Creating new issue ${templated.title}`);
  try {
    const issue = await tools.github.issues.create({
      ...repo,
      ...templated,
      assignees: assignees
        ? listToArray(assignees)
        : listToArray(attributes.assignees),
      labels: listToArray(attributes.labels),
      milestone:
        Number(tools.inputs.milestone || attributes.milestone) || undefined,
    });

    setOutputs(tools, issue.data);
    tools.log.success(
      `Created issue ${issue.data.title}#${issue.data.number}: ${issue.data.html_url}`
    );
  } catch (err: any) {
    return logError(tools, template, "creating", err);
  }
}
