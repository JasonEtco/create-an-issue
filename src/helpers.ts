import { Toolkit } from "actions-toolkit";
import nunjucks from "nunjucks";
import { z } from "zod";

export const frontmatterSchema = z
  .object({
    title: z.string(),
    assignees: z.union([z.array(z.string()), z.string()]).optional(),
    labels: z.union([z.array(z.string()), z.string()]).optional(),
    milestone: z.union([z.string(), z.number()]).optional(),
    name: z.string().optional(),
    about: z.string().optional(),
  })
  .strict();

export type FrontMatterAttributes = z.infer<typeof frontmatterSchema>;

export function setOutputs(
  tools: Toolkit,
  issue: { number: number; html_url: string }
) {
  tools.outputs.number = String(issue.number);
  tools.outputs.url = issue.html_url;
}

export function listToArray(
  list: string[] | string,
  env: nunjucks.Environment,
  context: object
) {
  if (!list) return [];
  const array = Array.isArray(list) ? list : list.split(", ");
  return array.map((item) => env.renderString(item, context));
}
