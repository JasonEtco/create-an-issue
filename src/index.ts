import { Toolkit } from "actions-toolkit";
import { createAnIssue } from "./action";

Toolkit.run(createAnIssue);
