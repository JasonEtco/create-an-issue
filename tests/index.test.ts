import nock from "nock";
import * as core from "@actions/core";
import { Toolkit } from "actions-toolkit";
import { Signale } from "signale";
import { createAnIssue } from "../src/action";

function generateToolkit() {
  const tools = new Toolkit({
    logger: new Signale({ disabled: true }),
  });

  jest.spyOn(tools.log, "info");
  jest.spyOn(tools.log, "error");
  jest.spyOn(tools.log, "success");

  // Turn core.setOutput into a mocked noop
  jest.spyOn(core, "setOutput").mockImplementation(() => {});

  // Turn core.setFailed into a mocked noop
  jest.spyOn(core, "setFailed").mockImplementation(() => {});

  tools.exit.success = jest.fn() as any;
  tools.exit.failure = jest.fn() as any;

  return tools;
}

describe("create-an-issue", () => {
  let tools: Toolkit;
  let params: any;

  beforeEach(() => {
    nock("https://api.github.com")
      .post(/\/repos\/.*\/.*\/issues/)
      .reply(200, (_, body: any) => {
        params = body;
        return {
          title: body.title,
          number: 1,
          html_url: "www",
        };
      });

    tools = generateToolkit();

    // Ensure that the filename input isn't set at the start of a test
    delete process.env.INPUT_FILENAME;

    // Simulate an environment variable added for the action
    process.env.EXAMPLE = "foo";
  });

  it("creates a new issue", async () => {
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();

    // Verify that the outputs were set
    expect(core.setOutput).toHaveBeenCalledTimes(2);
    expect(core.setOutput).toHaveBeenCalledWith("url", "www");
    expect(core.setOutput).toHaveBeenCalledWith("number", "1");
  });

  it("creates a new issue from a different template", async () => {
    process.env.INPUT_FILENAME = ".github/different-template.md";
    tools.context.payload = {
      repository: { owner: { login: "JasonEtco" }, name: "waddup" },
    };
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with some template variables", async () => {
    process.env.INPUT_FILENAME = ".github/variables.md";
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with the context.repo template variables", async () => {
    process.env.INPUT_FILENAME = ".github/context-repo-template.md";
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with assignees, labels and a milestone", async () => {
    process.env.INPUT_FILENAME = ".github/kitchen-sink.md";
    tools.context.payload = {
      repository: { owner: { login: "fieldsfarmer" }, name: "waddup" },
    };
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with assignees and labels as comma-delimited strings", async () => {
    process.env.INPUT_FILENAME = ".github/split-strings.md";
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with an assignee passed by input", async () => {
    process.env.INPUT_ASSIGNEES = "octocat";
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with multiple assignees passed by input", async () => {
    process.env.INPUT_ASSIGNEES = "octocat, JasonEtco";
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.success).toHaveBeenCalled();
    expect((tools.log.success as any).mock.calls).toMatchSnapshot();
  });

  it("creates a new issue with a milestone passed by input", async () => {
    process.env.INPUT_MILESTONE = "1";
    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(params.milestone).toBe(1);
    expect(tools.log.success).toHaveBeenCalled();
  });

  it("creates a new issue when updating existing issues is enabled but no issues with the same title exist", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .reply(200, {
        items: [],
      })
      .post(/\/repos\/.*\/.*\/issues/)
      .reply(200, (_, body: any) => {
        params = body;
        return {
          title: body.title,
          number: 1,
          html_url: "www",
        };
      });

    process.env.INPUT_UPDATE_EXISTING = "true";

    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.log.info).toHaveBeenCalledWith(
      "No existing issue found to update"
    );
    expect(tools.log.success).toHaveBeenCalled();
  });

  it("updates an existing open issue with the same title", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .query((parsedQuery) => {
        const q = parsedQuery["q"];
        if (typeof q === "string") {
          const args = q.split(" ");
          return (
            (args.includes("is:open") || args.includes("is:closed")) &&
            args.includes("is:issue")
          );
        } else {
          return false;
        }
      })
      .reply(200, {
        items: [{ number: 1, title: "Hello!" }],
      })
      .patch(/\/repos\/.*\/.*\/issues\/.*/)
      .reply(200, {});

    process.env.INPUT_UPDATE_EXISTING = "true";

    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.exit.success).toHaveBeenCalled();
  });

  it("escapes quotes in the search query", async () => {
    process.env.INPUT_FILENAME = ".github/quotes-in-title.md";

    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .query((parsedQuery) => {
        const q = parsedQuery["q"] as string;
        return q.includes('"This title \\"has quotes\\""');
      })
      .reply(200, {
        items: [{ number: 1, title: "Hello!" }],
      })
      .post(/\/repos\/.*\/.*\/issues/)
      .reply(200, {});

    await createAnIssue(tools);
    expect(tools.log.success).toHaveBeenCalled();
  });

  it("checks the value of update_existing", async () => {
    process.env.INPUT_UPDATE_EXISTING = "invalid";

    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.exit.failure).toHaveBeenCalledWith(
      "Invalid value update_existing=invalid, must be one of true or false"
    );
  });

  it("updates an existing closed issue with the same title", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .query((parsedQuery) => {
        const q = parsedQuery["q"];
        if (typeof q === "string") {
          const args = q.split(" ");
          return !args.includes("is:all") && args.includes("is:issue");
        } else {
          return false;
        }
      })
      .reply(200, {
        items: [{ number: 1, title: "Hello!", html_url: "/issues/1" }],
      })
      .patch(/\/repos\/.*\/.*\/issues\/.*/)
      .reply(200, {});

    process.env.INPUT_UPDATE_EXISTING = "true";
    process.env.INPUT_SEARCH_EXISTING = "all";

    await createAnIssue(tools);
    expect(tools.exit.success).toHaveBeenCalledWith(
      "Updated issue Hello!#1: /issues/1"
    );
  });

  it("finds, but does not update an existing issue with the same title", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .reply(200, {
        items: [{ number: 1, title: "Hello!", html_url: "/issues/1" }],
      });
    process.env.INPUT_UPDATE_EXISTING = "false";

    await createAnIssue(tools);
    expect(params).toMatchSnapshot();
    expect(tools.exit.success).toHaveBeenCalledWith(
      "Existing issue Hello!#1: /issues/1 found but not updated"
    );
  });

  it("exits when updating an issue fails", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .reply(200, {
        items: [{ number: 1, title: "Hello!", html_url: "/issues/1" }],
      })
      .patch(/\/repos\/.*\/.*\/issues\/.*/)
      .reply(500, {
        message: "Updating issue failed",
      });

    await createAnIssue(tools);
    expect(tools.exit.failure).toHaveBeenCalled();
  });

  it("logs a helpful error if creating an issue throws an error", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .reply(200, { items: [] })
      .post(/\/repos\/.*\/.*\/issues/)
      .reply(500, {
        message: "Validation error",
      });

    await createAnIssue(tools);
    expect(tools.log.error).toHaveBeenCalled();
    expect((tools.log.error as any).mock.calls).toMatchSnapshot();
    expect(tools.exit.failure).toHaveBeenCalled();
  });

  it("logs a helpful error if creating an issue throws an error with more errors", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .reply(200, { items: [] })
      .post(/\/repos\/.*\/.*\/issues/)
      .reply(500, {
        message: "Validation error",
        errors: [{ foo: true }],
      });

    await createAnIssue(tools);
    expect(tools.log.error).toHaveBeenCalled();
    expect((tools.log.error as any).mock.calls).toMatchSnapshot();
    expect(tools.exit.failure).toHaveBeenCalled();
  });

  it("logs a helpful error if updating an issue throws an error with more errors", async () => {
    nock.cleanAll();
    nock("https://api.github.com")
      .get(/\/search\/issues.*/)
      .reply(200, { items: [{ number: 1, title: "Hello!" }] })
      .patch(/\/repos\/.*\/.*\/issues\/.*/)
      .reply(500, {
        message: "Validation error",
        errors: [{ foo: true }],
      });

    process.env.INPUT_UPDATE_EXISTING = "true";

    await createAnIssue(tools);
    expect(tools.log.error).toHaveBeenCalled();
    expect((tools.log.error as any).mock.calls).toMatchSnapshot();
    expect(tools.exit.failure).toHaveBeenCalled();
  });

  it("logs a helpful error if the frontmatter is invalid", async () => {
    process.env.INPUT_FILENAME = ".github/invalid-frontmatter.md";

    await createAnIssue(tools);
    expect(tools.log.error).toHaveBeenCalled();
    expect((tools.log.error as any).mock.calls).toMatchSnapshot();
    expect(tools.exit.failure).toHaveBeenCalled();
  });
});
