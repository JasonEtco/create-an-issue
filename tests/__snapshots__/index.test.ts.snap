// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-an-issue checks the value of update_existing 1`] = `
{
  "assignees": [
    "octocat",
    "JasonEtco",
  ],
  "body": "Goodbye!",
  "labels": [],
  "milestone": 1,
  "title": "Hello!",
}
`;

exports[`create-an-issue creates a new issue 1`] = `
{
  "assignees": [],
  "body": "Goodbye!",
  "labels": [],
  "title": "Hello!",
}
`;

exports[`create-an-issue creates a new issue 2`] = `
[
  [
    "Created issue Hello!#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue from a different template 1`] = `
{
  "assignees": [],
  "body": "Goodbye!",
  "labels": [],
  "title": "Different file",
}
`;

exports[`create-an-issue creates a new issue from a different template 2`] = `
[
  [
    "Created issue Different file#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue when updating existing issues is enabled but no issues with the same title exist 1`] = `
{
  "assignees": [
    "octocat",
    "JasonEtco",
  ],
  "body": "Goodbye!",
  "labels": [],
  "milestone": 1,
  "title": "Hello!",
}
`;

exports[`create-an-issue creates a new issue with a milestone passed by input 1`] = `
{
  "assignees": [
    "octocat",
    "JasonEtco",
  ],
  "body": "Goodbye!",
  "labels": [],
  "milestone": 1,
  "title": "Hello!",
}
`;

exports[`create-an-issue creates a new issue with an assignee passed by input 1`] = `
{
  "assignees": [
    "octocat",
  ],
  "body": "Goodbye!",
  "labels": [],
  "title": "Hello!",
}
`;

exports[`create-an-issue creates a new issue with an assignee passed by input 2`] = `
[
  [
    "Created issue Hello!#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue with assignees and labels as comma-delimited strings 1`] = `
{
  "assignees": [
    "JasonEtco",
    "matchai",
  ],
  "body": "The action create-an-issue is the best action.",
  "labels": [
    "bug",
    "enhancement",
  ],
  "title": "DO EVERYTHING",
}
`;

exports[`create-an-issue creates a new issue with assignees and labels as comma-delimited strings 2`] = `
[
  [
    "Created issue DO EVERYTHING#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue with assignees, labels and a milestone 1`] = `
{
  "assignees": [
    "JasonEtco",
  ],
  "body": "The action create-an-issue is the best action.",
  "labels": [
    "bugs",
    "fieldsfarmer",
  ],
  "milestone": 2,
  "title": "DO EVERYTHING",
}
`;

exports[`create-an-issue creates a new issue with assignees, labels and a milestone 2`] = `
[
  [
    "Created issue DO EVERYTHING#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue with multiple assignees passed by input 1`] = `
{
  "assignees": [
    "octocat",
    "JasonEtco",
  ],
  "body": "Goodbye!",
  "labels": [],
  "title": "Hello!",
}
`;

exports[`create-an-issue creates a new issue with multiple assignees passed by input 2`] = `
[
  [
    "Created issue Hello!#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue with some template variables 1`] = `
{
  "assignees": [],
  "body": "The action create-an-issue is the best action.

Environment variable foo is great.
",
  "labels": [],
  "title": "Hello create-an-issue",
}
`;

exports[`create-an-issue creates a new issue with some template variables 2`] = `
[
  [
    "Created issue Hello create-an-issue#1: www",
  ],
]
`;

exports[`create-an-issue creates a new issue with the context.repo template variables 1`] = `
{
  "assignees": [],
  "body": "Are you looking for results on our gh-pages branch?

Just click this link! https://JasonEtco.github.io/waddup/path/to/my/pageindex.html
",
  "labels": [],
  "title": "This Issue has a generalizable GitHub Pages Link",
}
`;

exports[`create-an-issue creates a new issue with the context.repo template variables 2`] = `
[
  [
    "Created issue This Issue has a generalizable GitHub Pages Link#1: www",
  ],
]
`;

exports[`create-an-issue finds, but does not update an existing issue with the same title 1`] = `
{
  "assignees": [
    "octocat",
    "JasonEtco",
  ],
  "body": "Goodbye!",
  "labels": [],
  "milestone": 1,
  "title": "Hello!",
}
`;

exports[`create-an-issue logs a helpful error if creating an issue throws an error 1`] = `
[
  [
    "An error occurred while creating the issue. This might be caused by a malformed issue title, or a typo in the labels or assignees. Check .github/ISSUE_TEMPLATE.md!",
  ],
  [
    [HttpError: Validation error],
  ],
]
`;

exports[`create-an-issue logs a helpful error if creating an issue throws an error with more errors 1`] = `
[
  [
    "An error occurred while creating the issue. This might be caused by a malformed issue title, or a typo in the labels or assignees. Check .github/ISSUE_TEMPLATE.md!",
  ],
  [
    [HttpError: Validation error: {"foo":true}],
  ],
]
`;

exports[`create-an-issue logs a helpful error if the frontmatter is invalid 1`] = `
[
  [
    "An error occurred while parsing the issue. This might be caused by a malformed issue title, or a typo in the labels or assignees. Check .github/invalid-frontmatter.md!",
  ],
  [
    {
      "_errors": [
        "Unrecognized key(s) in object: 'not_a_thing'",
      ],
      "labels": {
        "_errors": [
          "Expected array, received number",
          "Expected string, received number",
        ],
      },
      "title": {
        "_errors": [
          "Required",
        ],
      },
    },
  ],
]
`;

exports[`create-an-issue logs a helpful error if updating an issue throws an error with more errors 1`] = `
[
  [
    "An error occurred while updating the issue. This might be caused by a malformed issue title, or a typo in the labels or assignees. Check .github/ISSUE_TEMPLATE.md!",
  ],
  [
    [HttpError: Validation error: {"foo":true}],
  ],
]
`;

exports[`create-an-issue updates an existing open issue with the same title 1`] = `
{
  "assignees": [
    "octocat",
    "JasonEtco",
  ],
  "body": "Goodbye!",
  "labels": [],
  "milestone": 1,
  "title": "Hello!",
}
`;
